alf/algorithms/generator.py

```diff
@@ -119,47 +119,102 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
-    optimizing a downstream objective Jx - eps = 0. 
-    
-    If using ``svgd3``, then the eps quantity represents the kernel grad
-        :math:`\nabla_{z'}k(z', z)`
+
+    Maintain an encoding network that takes (z, vec) as input and predicts a 
+    matrix-vector product (mvp) of the form :math:`y=J^{-1}(z)*vec`, where 
+    :math:`J^{-1}(z)` is the inverse of the Jacobian matrix of some function
+    :math:`f(z)`, and ``vec`` is a vector. This network is used in GPVI in 
+    computing the ``functional_gradient`` of the generator, where :math:`J^{-1}`
+    is the inverse of the Jacobian of the generator function w.r.t. input noise
+    :math:`z'`, and ``vec`` is the gradient of the kernel :math:`\nabla_{z'}(z', z)`.
+
+    Training of this network is done outside of the algorithm, where the network is 
+    trained to predict :math:`y` that minimize the  objective :math:`||Jy - vec||^2.
     """
 
     def __init__(self,
+                 input_dim,
                  output_dim,
-                 net: Network,
+                 hidden_size=100,
+                 num_hidden_layers=1,
                  optimizer=None,
                  name="InverseMVPAlgorithm"):
         r"""Create a InverseMVPAlgorithm.
         Args:
-            output_dim (int): total output dimension of the inverse_mvp net, will differ
-                between ``svgd3`` and ``minmax`` methods
-            net (Network): network for predicting outputs from inputs.
+            input_dim (int): dimension of input z
+            output_dim (int): output dimension, i.e., dimension of the mvp 
+            hidden_size (int): width of hidden layers
+            num_hidden_layers (int): number of hidden layers after 
             optimizer (torch.optim.Optimizer): (optional) optimizer for training.
             name (str): name of this Algorithm.
         """
+        assert input_dim <= output_dim
+
         if optimizer is None:
             optimizer = alf.optimizers.Adam(lr=1e-3)
         super().__init__(train_state_spec=(), optimizer=optimizer, name=name)
 
-        self._output_dim = output_dim
-        self._net = net
+        kernel_initializer = functools.partial(
+            alf.initializers.variance_scaling_init,
+            gain=1.0 / 2.0,
+            mode='fan_in',
+            distribution='truncated_normal',
+            nonlinearity=math_ops.identity)
+
+        self._z_dim = input_dim
+        self._vec_dim = output_dim
+        z_spec = TensorSpec(shape=(self._z_dim, ))
+        vec_spec = TensorSpec(shape=(self._vec_dim, ))
+
+        self._net = EncodingNetwork(
+            (z_spec, vec_spec),
+            input_preprocessors=(torch.nn.Linear(self._z_dim, hidden_size),
+                                 torch.nn.Linear(self._vec_dim, hidden_size)),
+            preprocessing_combiner=alf.layers.NestConcat(),
+            fc_layer_params=(2 * hidden_size, ) * num_hidden_layers,
+            activation=torch.relu_,
```

Comment:
Maybe we should allow `activation` as an input argument. Not sure about this though.

Reply:
Agreed, updated.